src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -2848,6 +2839,24 @@ public void visitThrow(JCThrow tree) {
         public void visitApply(JCMethodInvocation tree) {
             scanExpr(tree.meth);
             scanExprs(tree.args);
+
+            // If super(): at this point all initialization blocks will execute
+            Name name = TreeInfo.name(tree.meth);
+            if (isConstructor && name == names._super) {
+                forEachInitializer(classDef, false, def -> {
+                    scan(def);
+                    clearPendingExits(false);
+                });
+            }
+
+            // If this(): at this point all final uninitialized fields will get initialized
+            if (isConstructor && name == names._this) {
+                for (int address = firstadr; address < nextadr; address++) {
+                    VarSymbol sym = vardecls[address].sym;
+                    if (isFinalUninitializedField(sym) && !sym.isStatic())
+                        letInit(tree.pos(), sym);
+                }
+            }
```

Comment:
The common `isConstructor` check can be moved to an enclosing `if` statement:

```suggestion
            if (isConstructor) {
                // If super(): at this point all initialization blocks will execute
                Name name = TreeInfo.name(tree.meth);
                if (name == names._super) {
                    forEachInitializer(classDef, false, def -> {
                        scan(def);
                        clearPendingExits(false);
                    });
                }

                // If this(): at this point all final uninitialized fields will get initialized
                else if (name == names._this) {
                    for (int address = firstadr; address < nextadr; address++) {
                        VarSymbol sym = vardecls[address].sym;
                        if (isFinalUninitializedField(sym) && !sym.isStatic())
                            letInit(tree.pos(), sym);
                    }
                }
            }
```

Reply:
Yep, thanks - applied in 0e638da2daa.